<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Welcome to Pyllelic’s documentation! &mdash; Pyllelic 0.2.8 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="pyllelic package" href="pyllelic.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="#" class="icon icon-home"> Pyllelic
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Welcome to Pyllelic’s documentation!</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#quickstart">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="#demo-gif">Demo gif</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#dependencies-and-installation">Dependencies and Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#using-conda-preferred">Using Conda (preferred)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#docker-container">Docker container</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pypi-installation">PyPi installation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#id1">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#example-exploratory-use-in-jupyter-notebook">Example exploratory use in jupyter notebook</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#function-reference">Function Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html">pyllelic package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="pyllelic.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="pyllelic.html#module-pyllelic.config">pyllelic.config module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pyllelic.html#module-pyllelic.process">pyllelic.process module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pyllelic.html#module-pyllelic.pyllelic">pyllelic.pyllelic module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pyllelic.html#module-pyllelic.quma">pyllelic.quma module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pyllelic.html#module-pyllelic.visualization">pyllelic.visualization module</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#authors">Authors</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyllelic.html">pyllelic package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.config">pyllelic.config module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.process">pyllelic.process module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.pyllelic">pyllelic.pyllelic module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.quma">pyllelic.quma module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.visualization">pyllelic.visualization module</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Pyllelic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
      <li>Welcome to Pyllelic’s documentation!</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="welcome-to-pyllelic-s-documentation">
<h1>Welcome to Pyllelic’s documentation!<a class="headerlink" href="#welcome-to-pyllelic-s-documentation" title="Permalink to this headline"></a></h1>
<img src="https://github.com/Paradoxdruid/pyllelic/blob/master/assets/pyllelic_logo.png?raw=true" width="50" height="50" style="float: left; margin-right: 10px;"><p><a class="reference external" href="https://github.com/Paradoxdruid/pyllelic">pyllelic</a>: a tool for detection of allelic-specific methylation variation in bisulfite DNA sequencing files.</p>
<div class="toctree-wrapper compound">
</div>
<section id="quickstart">
<h2>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline"></a></h2>
<p>Run an interactive sample pyllelic environment in your web browser using <a class="reference external" href="https://mybinder.org">mybinder.org</a>:</p>
<a href="https://mybinder.org/v2/gh/Paradoxdruid/pyllelic/HEAD?filepath=notebooks/pyllelic_binder_example.ipynb"><img src="https://mybinder.org/badge_logo.svg"></a></section>
<section id="demo-gif">
<h2>Demo gif<a class="headerlink" href="#demo-gif" title="Permalink to this headline"></a></h2>
<p align="center">
<img src="https://github.com/Paradoxdruid/pyllelic/blob/master/assets/pyllelic_demo.gif?raw=true" alt="pyllelic demo gif">
</p></section>
</section>
<section id="dependencies-and-installation">
<h1>Dependencies and Installation<a class="headerlink" href="#dependencies-and-installation" title="Permalink to this headline"></a></h1>
<p>Source code for pyllelic is available on <a class="reference external" href="https://github.com/Paradoxdruid/pyllelic">Github</a></p>
<section id="using-conda-preferred">
<h2>Using Conda (preferred)<a class="headerlink" href="#using-conda-preferred" title="Permalink to this headline"></a></h2>
<p>Create a new conda environment using python 3.8:</p>
<p>Easiest:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get environment.yml file from this repo</span>
curl -L https://github.com/Paradoxdruid/pyllelic/blob/master/environment.yml?raw<span class="o">=</span><span class="nb">true</span> &gt; env.yml

<span class="c1"># Create and activate conda environment</span>
conda env create --file<span class="o">=</span>env.yml
conda activate pyllelic
</pre></div>
</div>
<details><summary><p>or more explict step by step instructions</p>
</summary><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda create --name pyllelic <span class="nv">python</span><span class="o">=</span><span class="m">3</span>.8
conda activate pyllelic
conda config --env --add channels conda-forge
conda config --env --add channels bioconda
conda config --env --add channels paradoxdruid
conda install pyllelic

<span class="c1"># Optional but usual use case:</span>
conda install notebook jupyter_contrib_nbextensions ipywidgets
</pre></div>
</div>
</details></section>
<section id="docker-container">
<h2>Docker container<a class="headerlink" href="#docker-container" title="Permalink to this headline"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>docker pull ghcr.io/paradoxdruid/pyllelic:latest
</pre></div>
</div>
</section>
<section id="pypi-installation">
<h2>PyPi installation<a class="headerlink" href="#pypi-installation" title="Permalink to this headline"></a></h2>
<details><summary><p>PyPi instructions</p>
</summary><p>This will require independent installation of samtools, bowtie2, and
bismark packages.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># PyPi</span>
python3 -m pip install pyllelic
<span class="c1"># or Github</span>
python3 -m pip install git+https://github.com/Paradoxdruid/pyllelic.git
</pre></div>
</div>
</details></section>
</section>
<section id="id1">
<h1>Quickstart<a class="headerlink" href="#id1" title="Permalink to this headline"></a></h1>
<section id="example-exploratory-use-in-jupyter-notebook">
<h2>Example exploratory use in jupyter notebook<a class="headerlink" href="#example-exploratory-use-in-jupyter-notebook" title="Permalink to this headline"></a></h2>
<p>Set up files:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pyllelic</span> <span class="kn">import</span> <span class="n">process</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="c1"># Retrieve promoter genomic sequence of region to analyze</span>
<span class="n">process</span><span class="o">.</span><span class="n">retrieve_seq</span><span class="p">(</span><span class="s2">&quot;tert_genome.txt&quot;</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="s2">&quot;chr5&quot;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1293000</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">1296000</span><span class="p">)</span>

<span class="c1"># Download a reference genome and bisulfite sequencing data</span>
<span class="c1"># Genome data from, e.g. http://hgdownload.soe.ucsc.edu/goldenPath/hg19</span>
<span class="c1"># Fastq data from, e.g. http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/</span>
<span class="n">genome</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/</span><span class="si">{your_directory}</span><span class="s2">/</span><span class="si">{genome_file_directory}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">fastq</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/</span><span class="si">{your_directory}</span><span class="s2">/</span><span class="si">{your_fastq_file.fastq.gz}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Use bismark tool to prepare bisulfite genome and align fastq to bam file</span>
<span class="n">process</span><span class="o">.</span><span class="n">prepare_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="c1"># can optionally give path to bowtie2 if not in PATH</span>
<span class="n">process</span><span class="o">.</span><span class="n">bismark</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">fastq</span><span class="p">)</span>

<span class="c1"># Sort and index the resultant bam file</span>
<span class="n">bamfile</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/</span><span class="si">{your_directory}</span><span class="s2">/</span><span class="si">{bam_filename}</span><span class="s2">.bam&quot;</span><span class="p">)</span>
<span class="n">process</span><span class="o">.</span><span class="n">sort_bam</span><span class="p">(</span><span class="n">bamfile</span><span class="p">)</span>
<span class="n">process</span><span class="o">.</span><span class="n">index_bam</span><span class="p">(</span><span class="n">bamfile</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bamfile</span><span class="o">.</span><span class="n">stem</span><span class="si">}</span><span class="s2">_sorted.bam&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Run pyllelic:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pyllelic</span> <span class="kn">import</span> <span class="n">pyllelic</span>

<span class="n">config</span> <span class="o">=</span> <span class="n">pyllelic</span><span class="o">.</span><span class="n">configure</span><span class="p">(</span>  <span class="c1"># Specify file and directory locations</span>
    <span class="n">base_path</span><span class="o">=</span><span class="s2">&quot;/home/jovyan/assets/&quot;</span><span class="p">,</span>
    <span class="n">prom_file</span><span class="o">=</span><span class="s2">&quot;tert_genome.txt&quot;</span><span class="p">,</span>
    <span class="n">prom_start</span><span class="o">=</span><span class="s2">&quot;1293200&quot;</span><span class="p">,</span>
    <span class="n">prom_end</span><span class="o">=</span><span class="s2">&quot;1296000&quot;</span><span class="p">,</span>
    <span class="n">chrom</span><span class="o">=</span><span class="s2">&quot;5&quot;</span><span class="p">,</span>
    <span class="n">offset</span><span class="o">=</span><span class="mi">1293000</span><span class="p">,</span>  <span class="c1"># start position of retrieved promoter sequence</span>
    <span class="c1"># viz_backend=&quot;plotly&quot;,</span>
    <span class="c1"># fname_pattern=r&quot;^[a-zA-Z]+_([a-zA-Z0-9]+)_.+bam$&quot;,</span>
    <span class="c1"># test_dir=&quot;test&quot;,</span>
    <span class="c1"># results_dir=&quot;results&quot;,</span>
<span class="p">)</span>

<span class="n">files_set</span> <span class="o">=</span> <span class="n">pyllelic</span><span class="o">.</span><span class="n">make_list_of_bam_files</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>  <span class="c1"># finds bam files</span>

<span class="c1"># Run pyllelic; make take some time depending on number of bam files</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pyllelic</span><span class="o">.</span><span class="n">pyllelic</span><span class="p">(</span><span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span> <span class="n">files_set</span><span class="o">=</span><span class="n">files_set</span><span class="p">)</span>

<span class="n">positions</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">positions</span>

<span class="n">cell_types</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">cell_types</span>

<span class="n">means_df</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">means</span>  <span class="c1"># mean methylation of reads</span>

<span class="n">modes_df</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">modes</span>  <span class="c1"># mode methylation of reads</span>

<span class="n">diff_df</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">diffs</span>  <span class="c1"># difference mean - mode of reads</span>

<span class="n">individual_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">individual_data</span>  <span class="c1"># read methylation values</span>

<span class="n">data</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;output.xlsx&quot;</span><span class="p">)</span>  <span class="c1"># save methylation results</span>

<span class="n">data</span><span class="o">.</span><span class="n">save_pickle</span><span class="p">(</span><span class="s2">&quot;my_run.pickle&quot;</span><span class="p">)</span>  <span class="c1"># save data object for later analysis</span>

<span class="n">data</span><span class="o">.</span><span class="n">write_means_modes_diffs</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;Run1_&quot;</span><span class="p">)</span>  <span class="c1"># write output data files</span>

<span class="n">data</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="s2">&quot;CELL_LINE&quot;</span><span class="p">,</span> <span class="s2">&quot;POSITION&quot;</span><span class="p">)</span>  <span class="c1"># visualize data for a point</span>

<span class="n">data</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">min_values</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># methylation level heatmap</span>

<span class="n">data</span><span class="o">.</span><span class="n">reads_graph</span><span class="p">()</span>  <span class="c1"># individual methylated / unmethylated reads graph</span>

<span class="n">data</span><span class="o">.</span><span class="n">quma_results</span><span class="p">[</span><span class="s2">&quot;CELL_LINE&quot;</span><span class="p">]</span>  <span class="c1"># see summary data for a cell line</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="function-reference">
<h1>Function Reference<a class="headerlink" href="#function-reference" title="Permalink to this headline"></a></h1>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyllelic.html">pyllelic package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.config">pyllelic.config module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.process">pyllelic.process module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.pyllelic">pyllelic.pyllelic module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.quma">pyllelic.quma module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyllelic.html#module-pyllelic.visualization">pyllelic.visualization module</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="authors">
<h1>Authors<a class="headerlink" href="#authors" title="Permalink to this headline"></a></h1>
<p>This software is developed as academic software by <a class="reference external" href="https://github.com/Paradoxdruid">Dr. Andrew J.
Bonham</a> at the <a class="reference external" href="https://www.msudenver.edu">Metropolitan State
University of Denver</a>. It is licensed
under the GPL v3.0.</p>
<p>This software incorporates implementation from
<a class="reference external" href="http://quma.cdb.riken.jp">QUMA</a>, licensed under the GPL v3.0.</p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pyllelic.html" class="btn btn-neutral float-right" title="pyllelic package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Andrew J. Bonham.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>