Search.setIndex({"docnames": ["index", "modules", "pyllelic"], "filenames": ["index.rst", "modules.rst", "pyllelic.rst"], "titles": ["Welcome to Pyllelic\u2019s documentation!", "pyllelic package", "pyllelic package"], "terms": {"tool": [0, 2], "detect": [0, 2], "allel": [0, 2], "specif": [0, 2], "methyl": [0, 2], "variat": [0, 2], "bisulfit": [0, 2], "dna": [0, 2], "sequenc": [0, 2], "file": [0, 2], "run": [0, 2], "an": [0, 2], "interact": 0, "sampl": 0, "environ": 0, "your": 0, "web": 0, "browser": [0, 2], "mybind": 0, "org": 0, "sourc": [0, 2], "code": 0, "avail": [0, 2], "github": [0, 2], "creat": [0, 2], "new": 0, "python": [0, 2], "3": 0, "8": 0, "easiest": 0, "get": 0, "yml": 0, "from": [0, 2], "thi": [0, 2], "repo": 0, "curl": 0, "l": 0, "http": [0, 2], "com": [0, 2], "paradoxdruid": 0, "blob": 0, "master": [0, 2], "raw": 0, "true": 0, "env": 0, "activ": 0, "more": [0, 2], "explict": 0, "step": [0, 2], "instruct": 0, "name": [0, 2], "config": [0, 1], "add": 0, "channel": 0, "forg": 0, "bioconda": 0, "option": [0, 2], "usual": [0, 2], "case": 0, "jupyter_contrib_nbextens": 0, "ipywidget": 0, "pull": 0, "ghcr": 0, "io": 0, "latest": 0, "requir": 0, "independ": 0, "samtool": [0, 2], "bowtie2": [0, 2], "bismark": [0, 2], "packag": 0, "python3": 0, "m": [0, 2], "pip": 0, "git": 0, "set": [0, 2], "up": [0, 2], "import": 0, "process": [0, 1], "pathlib": [0, 2], "path": [0, 2], "retriev": [0, 2], "promot": [0, 2], "genom": [0, 2], "region": [0, 2], "analyz": [0, 2], "retrieve_seq": [0, 2], "tert_genom": 0, "txt": [0, 2], "chrom": [0, 2], "chr5": [0, 2], "start": [0, 2], "1293000": [0, 2], "end": [0, 2], "1296000": [0, 2], "download": 0, "data": [0, 2], "e": [0, 2], "g": [0, 2], "hgdownload": [0, 2], "soe": [0, 2], "ucsc": [0, 2], "edu": [0, 2], "goldenpath": [0, 2], "hg19": 0, "fastq": [0, 2], "cse": 0, "encodedcc": 0, "wgencodehaibmethylrrb": 0, "your_directori": 0, "genome_file_directori": 0, "your_fastq_fil": 0, "gz": [0, 2], "prepar": [0, 2], "align": [0, 2], "bam": [0, 2], "prepare_genom": [0, 2], "can": [0, 2], "give": 0, "sort": [0, 2], "index": [0, 2], "result": [0, 2], "bamfil": [0, 2], "bam_filenam": 0, "sort_bam": [0, 2], "index_bam": [0, 2], "parent": 0, "f": [0, 2], "stem": 0, "_sort": 0, "configur": [0, 2], "specifi": [0, 2], "directori": [0, 2], "locat": [0, 2], "base_path": [0, 2], "home": 0, "jovyan": 0, "asset": 0, "prom_fil": [0, 2], "prom_start": [0, 2], "1293200": [0, 2], "prom_end": [0, 2], "5": [0, 2], "offset": [0, 2], "posit": [0, 2], "viz_backend": [0, 2], "plotli": [0, 2], "fname_pattern": [0, 2], "r": 0, "za": [0, 2], "z": [0, 2], "_": [0, 2], "z0": [0, 2], "9": [0, 2], "test_dir": [0, 2], "test": [0, 2], "results_dir": [0, 2], "files_set": [0, 2], "make_list_of_bam_fil": [0, 2], "find": 0, "make": 0, "take": [0, 2], "some": [0, 2], "time": [0, 2], "number": [0, 2], "cell_typ": [0, 2], "means_df": 0, "mean": [0, 2], "read": [0, 2], "modes_df": 0, "mode": [0, 2], "diff_df": 0, "diff": [0, 2], "differ": [0, 2], "individual_data": [0, 2], "valu": [0, 2], "save": [0, 2], "output": [0, 2], "xlsx": [0, 2], "save_pickl": [0, 2], "my_run": 0, "pickl": [0, 2], "object": [0, 2], "later": 0, "analysi": [0, 2], "write_means_modes_diff": [0, 2], "filenam": [0, 2], "run1_": 0, "write": 0, "histogram": [0, 2], "cell_lin": [0, 2], "visual": [0, 1], "point": [0, 2], "heatmap": [0, 2], "min_valu": [0, 2], "1": [0, 2], "level": [0, 2], "reads_graph": [0, 2], "individu": [0, 2], "unmethyl": [0, 2], "graph": [0, 2], "quma_result": [0, 2], "see": 0, "summari": 0, "cell": [0, 2], "line": [0, 2], "submodul": [0, 1], "modul": [0, 1], "quma": [0, 1], "main": [0, 1], "softwar": 0, "develop": 0, "academ": 0, "dr": 0, "andrew": 0, "j": 0, "bonham": 0, "metropolitan": 0, "state": 0, "univers": 0, "denver": 0, "It": 0, "licens": 0, "under": 0, "gpl": 0, "v3": 0, "0": [0, 2], "incorpor": 0, "implement": 0, "class": 2, "base_directori": 2, "posixpath": 2, "promoter_fil": 2, "results_directori": 2, "analysis_directori": 2, "promoter_start": 2, "int": 2, "promoter_end": 2, "chromosom": 2, "str": 2, "1298163": 2, "pattern": 2, "re": 2, "compil": 2, "base": 2, "util": 2, "pre": 2, "us": 2, "except": 2, "filenameerror": 2, "error": 2, "invalid": 2, "filetyp": 2, "shellcommanderror": 2, "shell": 2, "aren": 2, "t": 2, "instal": 2, "helper": 2, "function": 2, "extern": 2, "document": 2, "felixkrueg": 2, "tree": 2, "doc": 2, "paramet": 2, "filepath": 2, "return": 2, "command": 2, "discard": 2, "type": 2, "rais": 2, "bowtie2_fastq_to_bam": 2, "core": 2, "build": 2, "bowti": 2, "convert": 2, "build_bowtie2_index": 2, "fasta": 2, "fastq_to_list": 2, "list": 2, "seqrecord": 2, "memori": 2, "record_list": 2, "intens": 2, "oper": 2, "biopython": 2, "record": 2, "bool": 2, "verif": 2, "make_records_to_dictionari": 2, "dict": 2, "dictionari": 2, "kei": 2, "methbank_bed_to_indiv_data": 2, "stop": 2, "viz": 2, "genomicpositiondata": 2, "methbank": 2, "bed": 2, "df": 2, "ngdc": 2, "cncb": 2, "ac": 2, "cn": 2, "format": 2, "mostli": 2, "complet": 2, "none": 2, "hg38": 2, "bigzip": 2, "unprocess": 2, "program": 2, "bismark_genome_prepar": 2, "interest": 2, "store": 2, "pysam": 2, "reduc": 2, "represent": 2, "bisulf": 2, "ad_stat": 2, "sig": 2, "stat": 2, "union": 2, "_supportsarrai": 2, "dtype": 2, "_nestedsequ": 2, "float": 2, "complex": 2, "byte": 2, "crit": 2, "namedtupl": 2, "anderson_darling_test": 2, "alia": 2, "field": 2, "2": 2, "bamoutput": 2, "sam_directori": 2, "genome_str": 2, "storag": 2, "contain": 2, "genome_valu": 2, "content": 2, "pd": 2, "given": 2, "when": 2, "initi": 2, "automat": 2, "perform": 2, "qumaresult": 2, "final": 2, "aggreg": 2, "metric": 2, "well": 2, "expos": 2, "method": 2, "plot": 2, "statist": 2, "allelic_data": 2, "datafram": 2, "chi": 2, "squar": 2, "p": 2, "minu": 2, "file_nam": 2, "static": 2, "from_pickl": 2, "back": 2, "generate_ad_stat": 2, "gener": 2, "anderson": 2, "darl": 2, "normal": 2, "d": 2, "width": 2, "800": 2, "height": 2, "2000": 2, "data_typ": 2, "backend": 2, "displai": 2, "figur": 2, "show": 2, "across": 2, "minimum": 2, "must": 2, "exist": 2, "default": 2, "all": 2, "pvalu": 2, "overrid": 2, "valueerror": 2, "fraction": 2, "site": 2, "No": 2, "unabl": 2, "than": 2, "20": 2, "onc": 2, "excel": 2, "sig_methylation_differ": 2, "bar": 2, "chart": 2, "significantli": 2, "subset": 2, "summarize_allelic_data": 2, "onli": 2, "like": 2, "wite": 2, "out": 2, "futur": 2, "desir": 2, "root": 2, "read_fil": 2, "genomic_fil": 2, "style": 2, "quma_output": 2, "our": 2, "environment": 2, "variabl": 2, "where": 2, "occur": 2, "put": 2, "sub": 2, "folder": 2, "genmic": 2, "ar": 2, "regex": 2, "which": 2, "dataclass": 2, "instanc": 2, "check": 2, "valid": 2, "wrapper": 2, "call": 2, "routin": 2, "quantifi": 2, "po": 2, "seq": 2, "wrap": 2, "gfile_cont": 2, "qfile_cont": 2, "parser": 2, "convers": 2, "refer": 2, "form": 2, "tabular": 2, "dir": 2, "gdir": 2, "exc": 2, "intermedi": 2, "includ": 2, "directon": 2, "direct": 2, "whether": 2, "meet": 2, "exclus": 2, "criteria": 2, "qali": 2, "gali": 2, "val": 2, "perc": 2, "pconv": 2, "gap": 2, "menum": 2, "unconv": 2, "conv": 2, "match": 2, "alimi": 2, "alilen": 2, "alig": 2, "comparison": 2, "_create_heatmap": 2, "title_typ": 2, "being": 2, "matplotlib": 2, "go": 2, "plt": 2, "_create_histogram": 2, "provid": 2, "_create_methylation_diffs_bar_graph": 2, "signific": 2, "_make_binari": 2, "_make_methyl_df": 2, "row": 2, "_make_stacked_fig": 2, "_make_stacked_mpl_fig": 2, "_make_stacked_plotly_fig": 2, "interfac": 2, "exampl": 2, "usag": 2, "o": 2, "my_data": 2, "fh_cellline_tissu": 2, "hg19chr5": 2, "chr": 2, "s": 2, "would": 2, "prefix": 2, "indic": 2, "__main__": 2, "run_pyllel": 2}, "objects": {"pyllelic": [[2, 0, 0, "-", "__main__"], [2, 0, 0, "-", "config"], [2, 0, 0, "-", "process"], [2, 0, 0, "-", "pyllelic"], [2, 0, 0, "-", "quma"], [2, 0, 0, "-", "visualization"]], "pyllelic.__main__": [[2, 1, 1, "", "run_pyllelic"]], "pyllelic.config": [[2, 2, 1, "", "Config"]], "pyllelic.config.Config": [[2, 3, 1, "", "analysis_directory"], [2, 3, 1, "", "base_directory"], [2, 3, 1, "", "chromosome"], [2, 3, 1, "", "fname_pattern"], [2, 3, 1, "", "offset"], [2, 3, 1, "", "promoter_end"], [2, 3, 1, "", "promoter_file"], [2, 3, 1, "", "promoter_start"], [2, 3, 1, "", "results_directory"], [2, 3, 1, "", "viz_backend"]], "pyllelic.process": [[2, 4, 1, "", "FileNameError"], [2, 4, 1, "", "ShellCommandError"], [2, 1, 1, "", "bismark"], [2, 1, 1, "", "bowtie2_fastq_to_bam"], [2, 1, 1, "", "build_bowtie2_index"], [2, 1, 1, "", "fastq_to_list"], [2, 1, 1, "", "index_bam"], [2, 1, 1, "", "make_records_to_dictionary"], [2, 1, 1, "", "methbank_bed_to_indiv_data"], [2, 1, 1, "", "prepare_genome"], [2, 1, 1, "", "retrieve_seq"], [2, 1, 1, "", "sort_bam"]], "pyllelic.pyllelic": [[2, 2, 1, "", "AD_stats"], [2, 2, 1, "", "BamOutput"], [2, 2, 1, "", "GenomicPositionData"], [2, 2, 1, "", "QumaResult"], [2, 1, 1, "", "configure"], [2, 1, 1, "", "make_list_of_bam_files"], [2, 1, 1, "", "pyllelic"]], "pyllelic.pyllelic.AD_stats": [[2, 3, 1, "", "crits"], [2, 3, 1, "", "sig"], [2, 3, 1, "", "stat"]], "pyllelic.pyllelic.BamOutput": [[2, 3, 1, "", "genome_values"], [2, 3, 1, "", "name"], [2, 3, 1, "", "positions"], [2, 3, 1, "", "values"]], "pyllelic.pyllelic.GenomicPositionData": [[2, 3, 1, "", "allelic_data"], [2, 3, 1, "", "cell_types"], [2, 3, 1, "", "config"], [2, 3, 1, "", "diffs"], [2, 3, 1, "", "file_names"], [2, 3, 1, "", "files_set"], [2, 5, 1, "", "from_pickle"], [2, 5, 1, "", "generate_ad_stats"], [2, 5, 1, "", "heatmap"], [2, 5, 1, "", "histogram"], [2, 3, 1, "", "individual_data"], [2, 3, 1, "", "means"], [2, 3, 1, "", "modes"], [2, 3, 1, "", "positions"], [2, 3, 1, "", "quma_results"], [2, 5, 1, "", "reads_graph"], [2, 5, 1, "", "save"], [2, 5, 1, "", "save_pickle"], [2, 5, 1, "", "sig_methylation_differences"], [2, 5, 1, "", "summarize_allelic_data"], [2, 5, 1, "", "write_means_modes_diffs"]], "pyllelic.pyllelic.QumaResult": [[2, 3, 1, "", "quma_output"], [2, 3, 1, "", "values"]], "pyllelic.quma": [[2, 2, 1, "", "Fasta"], [2, 2, 1, "", "Quma"], [2, 2, 1, "", "Reference"], [2, 2, 1, "", "Result"]], "pyllelic.quma.Fasta": [[2, 3, 1, "", "com"], [2, 3, 1, "", "pos"], [2, 3, 1, "", "seq"]], "pyllelic.quma.Quma": [[2, 3, 1, "", "data"], [2, 3, 1, "", "values"]], "pyllelic.quma.Reference": [[2, 3, 1, "", "dir"], [2, 3, 1, "", "exc"], [2, 3, 1, "", "fasta"], [2, 3, 1, "", "gdir"], [2, 3, 1, "", "res"]], "pyllelic.quma.Result": [[2, 3, 1, "", "aliLen"], [2, 3, 1, "", "aliMis"], [2, 3, 1, "", "conv"], [2, 3, 1, "", "gAli"], [2, 3, 1, "", "gap"], [2, 3, 1, "", "match"], [2, 3, 1, "", "menum"], [2, 3, 1, "", "pconv"], [2, 3, 1, "", "perc"], [2, 3, 1, "", "qAli"], [2, 3, 1, "", "unconv"], [2, 3, 1, "", "val"]], "pyllelic.visualization": [[2, 1, 1, "", "_create_heatmap"], [2, 1, 1, "", "_create_histogram"], [2, 1, 1, "", "_create_methylation_diffs_bar_graph"], [2, 1, 1, "", "_make_binary"], [2, 1, 1, "", "_make_methyl_df"], [2, 1, 1, "", "_make_stacked_fig"], [2, 1, 1, "", "_make_stacked_mpl_fig"], [2, 1, 1, "", "_make_stacked_plotly_fig"]]}, "objtypes": {"0": "py:module", "1": "py:function", "2": "py:class", "3": "py:attribute", "4": "py:exception", "5": "py:method"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"], "2": ["py", "class", "Python class"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "exception", "Python exception"], "5": ["py", "method", "Python method"]}, "titleterms": {"welcom": 0, "pyllel": [0, 1, 2], "s": 0, "document": 0, "quickstart": 0, "demo": 0, "gif": 0, "depend": 0, "instal": 0, "us": 0, "conda": 0, "prefer": 0, "docker": 0, "contain": 0, "pypi": 0, "exampl": 0, "exploratori": 0, "jupyt": 0, "notebook": 0, "function": 0, "refer": 0, "author": 0, "packag": [1, 2], "submodul": 2, "config": 2, "modul": 2, "process": 2, "quma": 2, "visual": 2, "main": 2}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 6, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 56}})