<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyllelic.pyllelic &mdash; Pyllelic 0.2.8 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> Pyllelic
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome to Pyllelicâ€™s documentation!</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../index.html#quickstart">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../index.html#demo-gif">Demo gif</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../index.html#dependencies-and-installation">Dependencies and Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../index.html#using-conda-preferred">Using Conda (preferred)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../index.html#docker-container">Docker container</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../index.html#pypi-installation">PyPi installation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../index.html#id1">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../index.html#example-exploratory-use-in-jupyter-notebook">Example exploratory use in jupyter notebook</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../index.html#function-reference">Function Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html">pyllelic package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.config">pyllelic.config module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config"><code class="docutils literal notranslate"><span class="pre">Config</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.process">pyllelic.process module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.FileNameError"><code class="docutils literal notranslate"><span class="pre">FileNameError</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.ShellCommandError"><code class="docutils literal notranslate"><span class="pre">ShellCommandError</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.bismark"><code class="docutils literal notranslate"><span class="pre">bismark()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.bowtie2_fastq_to_bam"><code class="docutils literal notranslate"><span class="pre">bowtie2_fastq_to_bam()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.build_bowtie2_index"><code class="docutils literal notranslate"><span class="pre">build_bowtie2_index()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.convert_methbank_bed"><code class="docutils literal notranslate"><span class="pre">convert_methbank_bed()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.fastq_to_list"><code class="docutils literal notranslate"><span class="pre">fastq_to_list()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.index_bam"><code class="docutils literal notranslate"><span class="pre">index_bam()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.make_records_to_dictionary"><code class="docutils literal notranslate"><span class="pre">make_records_to_dictionary()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.prepare_genome"><code class="docutils literal notranslate"><span class="pre">prepare_genome()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.retrieve_seq"><code class="docutils literal notranslate"><span class="pre">retrieve_seq()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.sort_bam"><code class="docutils literal notranslate"><span class="pre">sort_bam()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.pyllelic">pyllelic.pyllelic module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.AD_stats"><code class="docutils literal notranslate"><span class="pre">AD_stats</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.BamOutput"><code class="docutils literal notranslate"><span class="pre">BamOutput</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.QumaResult"><code class="docutils literal notranslate"><span class="pre">QumaResult</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.configure"><code class="docutils literal notranslate"><span class="pre">configure()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.make_list_of_bam_files"><code class="docutils literal notranslate"><span class="pre">make_list_of_bam_files()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.pyllelic"><code class="docutils literal notranslate"><span class="pre">pyllelic()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.quma">pyllelic.quma module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Fasta"><code class="docutils literal notranslate"><span class="pre">Fasta</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Quma"><code class="docutils literal notranslate"><span class="pre">Quma</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Reference"><code class="docutils literal notranslate"><span class="pre">Reference</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result"><code class="docutils literal notranslate"><span class="pre">Result</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.visualization">pyllelic.visualization module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._create_heatmap"><code class="docutils literal notranslate"><span class="pre">_create_heatmap()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._create_histogram"><code class="docutils literal notranslate"><span class="pre">_create_histogram()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._create_methylation_diffs_bar_graph"><code class="docutils literal notranslate"><span class="pre">_create_methylation_diffs_bar_graph()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_binary"><code class="docutils literal notranslate"><span class="pre">_make_binary()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_methyl_df"><code class="docutils literal notranslate"><span class="pre">_make_methyl_df()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_stacked_fig"><code class="docutils literal notranslate"><span class="pre">_make_stacked_fig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_stacked_mpl_fig"><code class="docutils literal notranslate"><span class="pre">_make_stacked_mpl_fig()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_stacked_plotly_fig"><code class="docutils literal notranslate"><span class="pre">_make_stacked_plotly_fig()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.__main__">pyllelic.main module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.__main__.run_pyllelic"><code class="docutils literal notranslate"><span class="pre">run_pyllelic()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../index.html#authors">Authors</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../pyllelic.html">pyllelic package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.config">pyllelic.config module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config"><code class="docutils literal notranslate"><span class="pre">Config</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.analysis_directory"><code class="docutils literal notranslate"><span class="pre">Config.analysis_directory</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.base_directory"><code class="docutils literal notranslate"><span class="pre">Config.base_directory</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.chromosome"><code class="docutils literal notranslate"><span class="pre">Config.chromosome</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.fname_pattern"><code class="docutils literal notranslate"><span class="pre">Config.fname_pattern</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.offset"><code class="docutils literal notranslate"><span class="pre">Config.offset</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.promoter_end"><code class="docutils literal notranslate"><span class="pre">Config.promoter_end</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.promoter_file"><code class="docutils literal notranslate"><span class="pre">Config.promoter_file</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.promoter_start"><code class="docutils literal notranslate"><span class="pre">Config.promoter_start</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.results_directory"><code class="docutils literal notranslate"><span class="pre">Config.results_directory</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.config.Config.viz_backend"><code class="docutils literal notranslate"><span class="pre">Config.viz_backend</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.process">pyllelic.process module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.FileNameError"><code class="docutils literal notranslate"><span class="pre">FileNameError</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.ShellCommandError"><code class="docutils literal notranslate"><span class="pre">ShellCommandError</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.bismark"><code class="docutils literal notranslate"><span class="pre">bismark()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.bowtie2_fastq_to_bam"><code class="docutils literal notranslate"><span class="pre">bowtie2_fastq_to_bam()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.build_bowtie2_index"><code class="docutils literal notranslate"><span class="pre">build_bowtie2_index()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.convert_methbank_bed"><code class="docutils literal notranslate"><span class="pre">convert_methbank_bed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.fastq_to_list"><code class="docutils literal notranslate"><span class="pre">fastq_to_list()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.index_bam"><code class="docutils literal notranslate"><span class="pre">index_bam()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.make_records_to_dictionary"><code class="docutils literal notranslate"><span class="pre">make_records_to_dictionary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.prepare_genome"><code class="docutils literal notranslate"><span class="pre">prepare_genome()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.retrieve_seq"><code class="docutils literal notranslate"><span class="pre">retrieve_seq()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.process.sort_bam"><code class="docutils literal notranslate"><span class="pre">sort_bam()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.pyllelic">pyllelic.pyllelic module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.AD_stats"><code class="docutils literal notranslate"><span class="pre">AD_stats</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.AD_stats.crits"><code class="docutils literal notranslate"><span class="pre">AD_stats.crits</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.AD_stats.sig"><code class="docutils literal notranslate"><span class="pre">AD_stats.sig</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.AD_stats.stat"><code class="docutils literal notranslate"><span class="pre">AD_stats.stat</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.BamOutput"><code class="docutils literal notranslate"><span class="pre">BamOutput</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.BamOutput.genome_values"><code class="docutils literal notranslate"><span class="pre">BamOutput.genome_values</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.BamOutput.name"><code class="docutils literal notranslate"><span class="pre">BamOutput.name</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.BamOutput.positions"><code class="docutils literal notranslate"><span class="pre">BamOutput.positions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.BamOutput.values"><code class="docutils literal notranslate"><span class="pre">BamOutput.values</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.allelic_data"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.allelic_data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.cell_types"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.cell_types</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.config"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.config</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.diffs"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.diffs</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.file_names"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.file_names</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.files_set"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.files_set</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.from_pickle"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.from_pickle()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.generate_ad_stats"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.generate_ad_stats()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.heatmap"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.heatmap()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.histogram"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.histogram()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.individual_data"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.individual_data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.means"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.means</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.modes"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.modes</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.positions"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.positions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.quma_results"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.quma_results</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.reads_graph"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.reads_graph()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.save"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.save()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.save_pickle"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.save_pickle()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.sig_methylation_differences"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.sig_methylation_differences()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.summarize_allelic_data"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.summarize_allelic_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.write_means_modes_diffs"><code class="docutils literal notranslate"><span class="pre">GenomicPositionData.write_means_modes_diffs()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.QumaResult"><code class="docutils literal notranslate"><span class="pre">QumaResult</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.QumaResult.quma_output"><code class="docutils literal notranslate"><span class="pre">QumaResult.quma_output</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.QumaResult.values"><code class="docutils literal notranslate"><span class="pre">QumaResult.values</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.configure"><code class="docutils literal notranslate"><span class="pre">configure()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.make_list_of_bam_files"><code class="docutils literal notranslate"><span class="pre">make_list_of_bam_files()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.pyllelic.pyllelic"><code class="docutils literal notranslate"><span class="pre">pyllelic()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.quma">pyllelic.quma module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Fasta"><code class="docutils literal notranslate"><span class="pre">Fasta</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Fasta.com"><code class="docutils literal notranslate"><span class="pre">Fasta.com</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Fasta.pos"><code class="docutils literal notranslate"><span class="pre">Fasta.pos</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Fasta.seq"><code class="docutils literal notranslate"><span class="pre">Fasta.seq</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Quma"><code class="docutils literal notranslate"><span class="pre">Quma</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Quma.data"><code class="docutils literal notranslate"><span class="pre">Quma.data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Quma.values"><code class="docutils literal notranslate"><span class="pre">Quma.values</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Reference"><code class="docutils literal notranslate"><span class="pre">Reference</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Reference.dir"><code class="docutils literal notranslate"><span class="pre">Reference.dir</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Reference.exc"><code class="docutils literal notranslate"><span class="pre">Reference.exc</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Reference.fasta"><code class="docutils literal notranslate"><span class="pre">Reference.fasta</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Reference.gdir"><code class="docutils literal notranslate"><span class="pre">Reference.gdir</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Reference.res"><code class="docutils literal notranslate"><span class="pre">Reference.res</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result"><code class="docutils literal notranslate"><span class="pre">Result</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.aliLen"><code class="docutils literal notranslate"><span class="pre">Result.aliLen</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.aliMis"><code class="docutils literal notranslate"><span class="pre">Result.aliMis</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.conv"><code class="docutils literal notranslate"><span class="pre">Result.conv</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.gAli"><code class="docutils literal notranslate"><span class="pre">Result.gAli</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.gap"><code class="docutils literal notranslate"><span class="pre">Result.gap</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.match"><code class="docutils literal notranslate"><span class="pre">Result.match</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.menum"><code class="docutils literal notranslate"><span class="pre">Result.menum</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.pconv"><code class="docutils literal notranslate"><span class="pre">Result.pconv</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.perc"><code class="docutils literal notranslate"><span class="pre">Result.perc</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.qAli"><code class="docutils literal notranslate"><span class="pre">Result.qAli</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.unconv"><code class="docutils literal notranslate"><span class="pre">Result.unconv</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../pyllelic.html#pyllelic.quma.Result.val"><code class="docutils literal notranslate"><span class="pre">Result.val</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.visualization">pyllelic.visualization module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._create_heatmap"><code class="docutils literal notranslate"><span class="pre">_create_heatmap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._create_histogram"><code class="docutils literal notranslate"><span class="pre">_create_histogram()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._create_methylation_diffs_bar_graph"><code class="docutils literal notranslate"><span class="pre">_create_methylation_diffs_bar_graph()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_binary"><code class="docutils literal notranslate"><span class="pre">_make_binary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_methyl_df"><code class="docutils literal notranslate"><span class="pre">_make_methyl_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_stacked_fig"><code class="docutils literal notranslate"><span class="pre">_make_stacked_fig()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_stacked_mpl_fig"><code class="docutils literal notranslate"><span class="pre">_make_stacked_mpl_fig()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.visualization._make_stacked_plotly_fig"><code class="docutils literal notranslate"><span class="pre">_make_stacked_plotly_fig()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../pyllelic.html#module-pyllelic.__main__">pyllelic.main module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pyllelic.html#pyllelic.__main__.run_pyllelic"><code class="docutils literal notranslate"><span class="pre">run_pyllelic()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Pyllelic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>pyllelic.pyllelic</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pyllelic.pyllelic</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="sd">&quot;&quot;&quot;pyllelic: a tool for detection of allelic-specific variation</span>
<span class="sd">   in reduced representation bisulfate DNA sequencing.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">import</span> <span class="nn">pickle</span>  <span class="c1"># nosec</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">signal</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span><span class="p">,</span> <span class="n">cpu_count</span>
<span class="kn">from</span> <span class="nn">multiprocessing.pool</span> <span class="kn">import</span> <span class="n">AsyncResult</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Match</span><span class="p">,</span> <span class="n">NamedTuple</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Pattern</span><span class="p">,</span> <span class="n">Tuple</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">from</span> <span class="nn">numpy.typing</span> <span class="kn">import</span> <span class="n">ArrayLike</span><span class="p">,</span> <span class="n">NDArray</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span> <span class="nn">tqdm.auto</span> <span class="kn">import</span> <span class="n">tqdm</span>

<span class="kn">from</span> <span class="nn">pyllelic</span> <span class="kn">import</span> <span class="n">quma</span>
<span class="kn">from</span> <span class="nn">pyllelic</span> <span class="kn">import</span> <span class="n">visualization</span> <span class="k">as</span> <span class="n">viz</span>
<span class="kn">from</span> <span class="nn">pyllelic.config</span> <span class="kn">import</span> <span class="n">Config</span>

<span class="c1"># Initialized multiprocessing limits</span>
<span class="n">NUM_THREADS</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>

<span class="c1"># SANITY ALIGNMENT LIMITS</span>
<span class="n">MIN_NUMBER_READS</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">MIN_PERCENT_ALIGNMENT</span> <span class="o">=</span> <span class="mi">50</span>


<div class="viewcode-block" id="AD_stats"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.AD_stats">[docs]</a><span class="k">class</span> <span class="nc">AD_stats</span><span class="p">(</span><span class="n">NamedTuple</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Helper class for NamedTuple results from anderson_darling_test&quot;&quot;&quot;</span>

    <span class="n">sig</span><span class="p">:</span> <span class="nb">bool</span>
    <span class="n">stat</span><span class="p">:</span> <span class="n">ArrayLike</span>
    <span class="n">crits</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">ArrayLike</span><span class="p">]</span></div>


<div class="viewcode-block" id="BamOutput"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.BamOutput">[docs]</a><span class="k">class</span> <span class="nc">BamOutput</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Storage container to process BAM sequencing files and store processed results.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sam_directory</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">genome_string</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">config</span><span class="p">:</span> <span class="n">Config</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_config</span><span class="p">:</span> <span class="n">Config</span> <span class="o">=</span> <span class="n">config</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">sam_directory</span><span class="p">)</span>
        <span class="sd">&quot;&quot;&quot;str: path to bam file analyzed.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="sd">&quot;&quot;&quot;Dict[str, str]: dictionary of reads at a given position&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">positions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_run_sam_and_extract_df</span><span class="p">(</span><span class="n">sam_directory</span><span class="p">)</span>
        <span class="sd">&quot;&quot;&quot;&quot;pd.Index: index of genomic positions in the bam file.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genome_values</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="sd">&quot;&quot;&quot;Dict[str, str]: dictionary of read files and contents.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_genome_parsing</span><span class="p">(</span><span class="n">genome_string</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_run_sam_and_extract_df</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sams</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Process samfiles, pulling out sequence and position data</span>
<span class="sd">        and writing to folders/files.</span>

<span class="sd">        Args:</span>
<span class="sd">            sams (Path): path to a samfile</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: list of unique positions in the samfile</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Grab the promoter region of interest</span>
        <span class="n">samm</span><span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">sams</span><span class="p">),</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
        <span class="n">itern</span> <span class="o">=</span> <span class="n">samm</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_config</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_config</span><span class="o">.</span><span class="n">promoter_start</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_config</span><span class="o">.</span><span class="n">promoter_end</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">position</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">sequence</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">itern</span><span class="p">:</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="n">position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
            <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">9</span><span class="p">])</span>

        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;positions&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence&quot;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="n">df2</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;positions&quot;</span><span class="p">)</span>
        <span class="n">df3</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span> <span class="o">=</span> <span class="n">df2</span><span class="o">.</span><span class="n">stack</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_write_bam_output</span><span class="p">(</span><span class="n">df2</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">df3</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">df2</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># type: ignore</span>

    <span class="k">def</span> <span class="nf">_write_bam_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">positions</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Index</span><span class="p">,</span> <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Extract alignments from sequencing reads and output text strings</span>
<span class="sd">        in bam values dictionary.</span>

<span class="sd">        Args:</span>
<span class="sd">            positions (pd.Index): list of unique positions</span>
<span class="sd">            df (pd.Series): series of sequencing reads</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">each1</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">:</span>
            <span class="n">alignments</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">each1</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

            <span class="n">read_file</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">each</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignments</span><span class="p">):</span>
                <span class="n">read_file</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;&gt;read&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">index</span><span class="p">)))</span>
                <span class="n">read_file</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">each</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">each1</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">read_file</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_genome_range</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">position</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">genome_string</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">offset</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Helper to return a genome string (e.g., &quot;ATCGACTAG&quot;)</span>
<span class="sd">        given a position and an entire string.</span>

<span class="sd">        Args:</span>
<span class="sd">            position: genomic position on chromesome</span>
<span class="sd">            genome_string: string representation of genomic promoter known sequence</span>
<span class="sd">            offset: genomic position of promoter to offset reads</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: genomic bases for indicated read / position</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: incorrect position</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">offset</span><span class="p">:</span>
            <span class="n">offset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_config</span><span class="o">.</span><span class="n">offset</span>

        <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">-</span> <span class="n">offset</span>
        <span class="n">end</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">-</span> <span class="n">offset</span><span class="p">)</span> <span class="o">+</span> <span class="mi">30</span>
        <span class="k">if</span> <span class="n">genome_string</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]:</span>
            <span class="k">return</span> <span class="n">genome_string</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Position not in genomic promoter file&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_genome_parsing</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome_string</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Writes out a list of genomic sequence strings for comparison to read data.</span>

<span class="sd">        Args:</span>
<span class="sd">            genome_string (str): genomic sequence</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Grab list of read files in that directory:</span>
        <span class="n">read_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="c1"># Now, process each file:</span>
        <span class="k">for</span> <span class="n">read_name</span> <span class="ow">in</span> <span class="n">read_files</span><span class="p">:</span>
            <span class="n">file_lines</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="c1"># Grab the genomic sequence</span>
            <span class="n">file_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;&gt;genome&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">read_name</span><span class="p">)))</span>
            <span class="n">file_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_range</span><span class="p">(</span><span class="n">read_name</span><span class="p">,</span> <span class="n">genome_string</span><span class="p">)))</span>

            <span class="c1"># Save the reads as a text file for each position</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome_values</span><span class="p">[</span><span class="n">read_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_lines</span><span class="p">)</span></div>


<div class="viewcode-block" id="QumaResult"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.QumaResult">[docs]</a><span class="k">class</span> <span class="nc">QumaResult</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Storage container to process and store quma-style methylation results.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">read_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">genomic_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">positions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_read_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">read_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genomic_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">genomic_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_positions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">positions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">quma_output</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">quma</span><span class="o">.</span><span class="n">Quma</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;List[quma.Quma]: list of Quma result objects.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pool_processing</span><span class="p">()</span>
        <span class="sd">&quot;&quot;&quot;pd.DataFrame: dataframe of quma methylation analysis values.&quot;&quot;&quot;</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_process_raw_quma</span><span class="p">(</span><span class="n">quma_data</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">quma</span><span class="o">.</span><span class="n">Reference</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Spit and process raw quma results into a pandas dataframe.</span>

<span class="sd">        Args:</span>
<span class="sd">            quma_data (List[quma.Reference]): list of quma References.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: intermediate dataframe to append</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">dots</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">res</span><span class="o">.</span><span class="n">res</span><span class="o">.</span><span class="n">val</span> <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">res</span><span class="o">.</span><span class="n">perc</span> <span class="o">&gt;</span> <span class="n">MIN_PERCENT_ALIGNMENT</span> <span class="k">else</span> <span class="s2">&quot;FAIL&quot;</span>
            <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">quma_data</span>
        <span class="p">]</span>

        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">dots</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;1&quot;</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">_pool_processing</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Helper to run quma in a multiprocessing pool.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: dataframe of quma results</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Set up a holding data frame from all the data</span>
        <span class="n">holding_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

        <span class="n">returns</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">AsyncResult</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">quma</span><span class="o">.</span><span class="n">Quma</span><span class="p">]]]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">NUM_THREADS</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_init_worker</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>

            <span class="k">for</span> <span class="n">position</span><span class="p">,</span> <span class="n">read</span><span class="p">,</span> <span class="n">genomic</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_positions</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_files</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genomic_files</span>
            <span class="p">):</span>

                <span class="n">result</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_thread_worker</span><span class="p">,</span>
                    <span class="p">(</span><span class="n">genomic</span><span class="p">,</span> <span class="n">read</span><span class="p">,</span> <span class="n">position</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="n">returns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

            <span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="n">result</span><span class="o">.</span><span class="n">get</span><span class="p">()</span> <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">returns</span><span class="p">]</span>

        <span class="c1"># Add to excel file</span>
        <span class="k">for</span> <span class="n">int_df</span><span class="p">,</span> <span class="n">quma_obj</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">holding_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">holding_df</span><span class="p">,</span> <span class="n">int_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">quma_output</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">quma_obj</span><span class="p">)</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">KeyError</span><span class="p">):</span>  <span class="c1"># pragma: no cover</span>
                <span class="k">pass</span>

        <span class="k">return</span> <span class="n">holding_df</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_init_worker</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Pool worker initializer for keyboard interrupt on Windows</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">signal</span><span class="o">.</span><span class="n">signal</span><span class="p">(</span><span class="n">signal</span><span class="o">.</span><span class="n">SIGINT</span><span class="p">,</span> <span class="n">signal</span><span class="o">.</span><span class="n">SIG_IGN</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_thread_worker</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">genomic_contents</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">read_contents</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">position</span><span class="p">:</span> <span class="nb">str</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">quma</span><span class="o">.</span><span class="n">Quma</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Queue worker for quma functions.</span>

<span class="sd">        Args:</span>
<span class="sd">            genomic_contents (str): genomic sequence</span>
<span class="sd">            read_contents (str): query sequence</span>
<span class="sd">            position(str): position of reads</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple[pd.DataFrame, quma.Quma]: dataframe of quma results and raw quma objs</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">quma_obj</span><span class="p">:</span> <span class="n">quma</span><span class="o">.</span><span class="n">Quma</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_access_quma</span><span class="p">(</span><span class="n">genomic_contents</span><span class="p">,</span> <span class="n">read_contents</span><span class="p">)</span>

        <span class="n">processed_quma</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_raw_quma</span><span class="p">(</span><span class="n">quma_obj</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="c1"># Next, add this readname to the holding data frame</span>
        <span class="n">int_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">position</span><span class="p">:</span> <span class="n">processed_quma</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">int_df</span><span class="p">,</span> <span class="n">quma_obj</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_access_quma</span><span class="p">(</span><span class="n">genomic_contents</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">read_contents</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">quma</span><span class="o">.</span><span class="n">Quma</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Helper function to run internal QUMA tool.</span>

<span class="sd">        Args:</span>
<span class="sd">            genomic_contents (str): genome sequence</span>
<span class="sd">            read_contents (str): query sequence</span>

<span class="sd">        Returns:</span>
<span class="sd">            quma.Quma: quma object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">quma</span><span class="o">.</span><span class="n">Quma</span><span class="p">(</span><span class="n">genomic_contents</span><span class="p">,</span> <span class="n">read_contents</span><span class="p">)</span></div>


<span class="c1">##################################################################################</span>


<div class="viewcode-block" id="GenomicPositionData"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData">[docs]</a><span class="k">class</span> <span class="nc">GenomicPositionData</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Class to process reduced representation bisulfite methylation sequencing data.</span>

<span class="sd">    When initialized, GenomicPositionData reads sequencing file (.bam) locations from</span>
<span class="sd">    a config object, and then automatically performs alignment into BamOutput objects,</span>
<span class="sd">    and then performs methylation analysis, storing the results as QumaResults.</span>

<span class="sd">    Finally, the aggregate data is analyzed to create some aggregate metrics such as</span>
<span class="sd">    means, modes, and differences (diffs), as well as expose methods for plotting</span>
<span class="sd">    and statistical analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">config</span><span class="p">:</span> <span class="n">Config</span><span class="p">,</span> <span class="n">files_set</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">:</span> <span class="n">Config</span> <span class="o">=</span> <span class="n">config</span>
        <span class="sd">&quot;&quot;&quot;Config: pyllelic config object.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">files_set</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">files_set</span>
        <span class="sd">&quot;&quot;&quot;List[str]: list of bam files analyzed.&quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">promoter_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">genome_base</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
            <span class="n">genome_base_lines</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">genome_base</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome_string</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">genome_base_lines</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_bam_output</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BamOutput</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_index_and_fetch</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">positions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_positions</span><span class="p">()</span>
        <span class="sd">&quot;&quot;&quot;List[str]: list of genomic positions in the data.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">file_names</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_bam_output</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="sd">&quot;&quot;&quot;List[str]: list of bam filenames in the data.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">cell_types</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_retrieve_cell_names</span><span class="p">()</span>
        <span class="sd">&quot;&quot;&quot;List[str]: list of cell types in the data.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">quma_results</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">QumaResult</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_quma_full_threaded</span><span class="p">()</span>
        <span class="sd">&quot;&quot;&quot;Dict[str, QumaResult]: list of QumaResults.&quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Summary Statistics&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_individual_position_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span>
                <span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
            <span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_individual_position_df_dict</span><span class="p">()</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">means</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_means</span><span class="p">()</span>
            <span class="sd">&quot;&quot;&quot;pd.DataFrame: dataframe of mean methylation values.&quot;&quot;&quot;</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_modes</span><span class="p">()</span>
            <span class="sd">&quot;&quot;&quot;pd.DataFrame: dataframe of modes of methylation values.&quot;&quot;&quot;</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">diffs</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_find_diffs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">means</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="p">)</span>
            <span class="sd">&quot;&quot;&quot;pd.DataFrame: df of difference mean minus mode methylation values.&quot;&quot;&quot;</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">individual_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_return_individual_data</span><span class="p">()</span>
            <span class="sd">&quot;&quot;&quot;pd.DataFrame: dataframe of individual methylation values.&quot;&quot;&quot;</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">allelic_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generate_chisquared_test_df</span><span class="p">()</span>
            <span class="sd">&quot;&quot;&quot;pd.DataFrame: dataframe of Chi-squared p-values.&quot;&quot;&quot;</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_index_and_fetch</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BamOutput</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Wrapper to call processing of each sam file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict[str, BamOutput]: dictionary of bam outputs&quot;&quot;&quot;</span>

        <span class="n">sam_path</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">base_directory</span> <span class="o">/</span> <span class="s2">&quot;test&quot;</span> <span class="o">/</span> <span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">files_set</span>
        <span class="p">]</span>
        <span class="n">bam_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BamOutput</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">sams</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">sam_path</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Process BAM Files&quot;</span><span class="p">):</span>
            <span class="n">bam_dict</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">sams</span><span class="p">)]</span> <span class="o">=</span> <span class="n">BamOutput</span><span class="p">(</span><span class="n">sams</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_string</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">bam_dict</span>

    <span class="k">def</span> <span class="nf">_calculate_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Return sorted list of all positions analyzed.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: sorted list of all positions analyzed.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bam_output</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">each</span> <span class="ow">in</span> <span class="n">sublist</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">each</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">positions</span><span class="p">)))</span>

    <span class="k">def</span> <span class="nf">_retrieve_cell_names</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Retrieve shortened cell line names from the list of filenames.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: list of cell line names.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cells</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">each</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">file_names</span><span class="p">:</span>
            <span class="n">match</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">fname_pattern</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">each</span><span class="p">)</span><span class="o">.</span><span class="n">name</span>
            <span class="p">)</span>
            <span class="n">cell_line_name</span><span class="p">:</span> <span class="nb">str</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">match</span><span class="p">,</span> <span class="n">Match</span><span class="p">):</span>
                <span class="n">cell_line_name</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># pragma: no cover</span>
                <span class="n">cell_line_name</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">each</span><span class="p">)</span><span class="o">.</span><span class="n">name</span>
            <span class="n">cells</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_line_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cells</span>

    <span class="k">def</span> <span class="nf">_quma_full_threaded</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">QumaResult</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Run external QUMA methylation analysis on all specified cell lines,</span>
<span class="sd">        using multiprocessing library.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict[str, QumaResult]: dictionary of cell_types and quma results</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">quma_results</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">QumaResult</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">bam_result</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_bam_output</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Process methylation&quot;</span>
        <span class="p">):</span>
            <span class="n">match</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">fname_pattern</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="o">.</span><span class="n">name</span>
            <span class="p">)</span>
            <span class="n">cell_line_name</span><span class="p">:</span> <span class="nb">str</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">match</span><span class="p">,</span> <span class="n">Match</span><span class="p">):</span>
                <span class="n">cell_line_name</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># pragma: no cover</span>
                <span class="n">cell_line_name</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="o">.</span><span class="n">name</span>

            <span class="n">read_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">each</span> <span class="k">for</span> <span class="n">each</span> <span class="ow">in</span> <span class="n">bam_result</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
            <span class="n">genomic_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">each</span> <span class="k">for</span> <span class="n">each</span> <span class="ow">in</span> <span class="n">bam_result</span><span class="o">.</span><span class="n">genome_values</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
            <span class="p">]</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="n">bam_result</span><span class="o">.</span><span class="n">positions</span>

            <span class="n">quma_results</span><span class="p">[</span><span class="n">cell_line_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">QumaResult</span><span class="p">(</span>
                <span class="n">read_files</span><span class="p">,</span> <span class="n">genomic_files</span><span class="p">,</span> <span class="n">positions</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">quma_results</span>

<div class="viewcode-block" id="GenomicPositionData.save"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.save">[docs]</a>    <span class="k">def</span> <span class="nf">save</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;output.xlsx&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Save quma results to an excel file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename (str): Filename to save to. Defaults to &quot;output.xlsx&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">base_directory</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">each</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">quma_results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">each</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="n">name</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicPositionData.save_pickle"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.save_pickle">[docs]</a>    <span class="k">def</span> <span class="nf">save_pickle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Save GenomicPositionData object as a pickled file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename (str): filename to save pickle</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output_file</span><span class="p">:</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicPositionData.from_pickle"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.from_pickle">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">from_pickle</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicPositionData&quot;</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Read pickled GenomicPositionData back to an object.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename (str): filename to read pickle</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicPositionData: GenomicPositionData object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">data</span><span class="p">:</span> <span class="n">GenomicPositionData</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>  <span class="c1"># nosec</span>

        <span class="k">return</span> <span class="n">data</span></div>

    <span class="k">def</span> <span class="nf">_create_individual_position_df_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Create a dictionary of invidual position data for use in aggregate dfs.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict[str, pd.DataFrame]: dictionary of invidual position data dfs</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">cell</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_return_individual_positions</span><span class="p">(</span><span class="n">cell</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cell_types</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="n">df_dict</span>

    <span class="k">def</span> <span class="nf">_process_means</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Process the mean values at each individual position for each cell line.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: dataframes of mean values for each position in each cell line.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_individual_position_data</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">df_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">df_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">df_dict</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
        <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="k">def</span> <span class="nf">_process_modes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Process the mode at each individual position for each cell line.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: dataframes of mode values for each individual position.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_individual_position_data</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">df_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">df_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">mode</span><span class="p">()</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">df_dict</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
        <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="k">def</span> <span class="nf">_return_individual_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Return a dataframe for methylation values at each pos for each cell line.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: methylation values for each position in each cell line</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_individual_position_data</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">df_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">df_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">stack</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
                <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
                <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
                <span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
            <span class="p">)</span>

        <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_dict</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">droplevel</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="k">def</span> <span class="nf">_generate_chisquared_test_df</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Return a dataframe for p-values values at each position for each cell line.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: Chi-squared p-values for each position in each cell line</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">individual_data</span><span class="o">.</span><span class="n">applymap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_chisquared_test</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_find_diffs</span><span class="p">(</span><span class="n">means_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">modes_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Find the differences between means and modes for each cell line at each pos</span>
<span class="sd">        in means and modes data.</span>

<span class="sd">        Args:</span>
<span class="sd">            means_df (pd.DataFrame): dataframe of means values</span>
<span class="sd">            modes_df (pd.DataFrame): dataframe of modes values</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: dataframe of difference values</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">means_df</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">modes_df</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_truncate_diffs</span><span class="p">(</span><span class="n">diffs_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Remove missing or non-interesting diffs, and sort by magnitude.</span>

<span class="sd">        Args:</span>
<span class="sd">            diffs_df (pd.DataFrame): dataframe of diffs values</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: truncated dataframe</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">diffs_df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="GenomicPositionData.write_means_modes_diffs"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.write_means_modes_diffs">[docs]</a>    <span class="k">def</span> <span class="nf">write_means_modes_diffs</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Wite out files of means, modes, and diffs for future analysis.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename (str): desired root filename</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">means</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">base_directory</span> <span class="o">/</span> <span class="nb">str</span><span class="p">(</span><span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot;_means.xlsx&quot;</span><span class="p">),</span>
            <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Means&quot;</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">base_directory</span> <span class="o">/</span> <span class="nb">str</span><span class="p">(</span><span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot;_modes.xlsx&quot;</span><span class="p">),</span>
            <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Modes&quot;</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">diffs</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">base_directory</span> <span class="o">/</span> <span class="nb">str</span><span class="p">(</span><span class="n">filename</span> <span class="o">+</span> <span class="s2">&quot;_diff.xlsx&quot;</span><span class="p">),</span>
            <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Diff&quot;</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span></div>

<div class="viewcode-block" id="GenomicPositionData.histogram"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.histogram">[docs]</a>    <span class="k">def</span> <span class="nf">histogram</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">cell_line</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">position</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">backend</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Display a graph figure showing fractional methylation in</span>
<span class="sd">        a given cell line at a given site.</span>

<span class="sd">        Args:</span>
<span class="sd">            cell_line (str): name of cell line</span>
<span class="sd">            position (str): genomic position</span>
<span class="sd">            backend (Optional[str]): plotting backend to override default</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: invalid plotting backend</span>
<span class="sd">            ValueError: No data available at that position</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">individual_data</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">backend</span><span class="p">:</span>
            <span class="n">backend</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">viz_backend</span>

        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cell_line</span><span class="p">,</span> <span class="n">position</span><span class="p">])):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No data available at that position&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;plotly&quot;</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">:</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_create_histogram</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">cell_line</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">backend</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>
        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;matplotlib&quot;</span><span class="p">:</span>
            <span class="n">_</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_create_histogram</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">cell_line</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">backend</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid plotting backend&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicPositionData.reads_graph"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.reads_graph">[docs]</a>    <span class="k">def</span> <span class="nf">reads_graph</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">cell_lines</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">backend</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Display a graph figure showing methylation of reads across cell lines.</span>

<span class="sd">        Args:</span>
<span class="sd">            cell_lines (Optional[List[str]]): set of cell lines to analyze,</span>
<span class="sd">                                            defaults to all cell lines.</span>
<span class="sd">            backend (Optional[str]): plotting backend to override default</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: invalid plotting backend</span>
<span class="sd">            ValueError: Unable to plot more than 20 cell lines at once.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">MAX_GRAPHS</span> <span class="o">=</span> <span class="mi">20</span>  <span class="c1"># Sanity limit</span>

        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">individual_data</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">backend</span><span class="p">:</span>
            <span class="n">backend</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">viz_backend</span>

        <span class="k">if</span> <span class="n">cell_lines</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cell_lines</span><span class="p">)]</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">MAX_GRAPHS</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unable to plot more than 20 cell lines at once.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;plotly&quot;</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">:</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_make_stacked_fig</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">backend</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>
        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;matplotlib&quot;</span><span class="p">:</span>
            <span class="n">_</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_make_stacked_fig</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">backend</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid plotting backend&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicPositionData.heatmap"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.heatmap">[docs]</a>    <span class="k">def</span> <span class="nf">heatmap</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">min_values</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">width</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">800</span><span class="p">,</span>
        <span class="n">height</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span>
        <span class="n">cell_lines</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;means&quot;</span><span class="p">,</span>
        <span class="n">backend</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Display a graph figure showing heatmap of mean methylation across</span>
<span class="sd">        cell lines.</span>

<span class="sd">        Args:</span>
<span class="sd">            min_values (int): minimum number of points data must exist at a position</span>
<span class="sd">            width (int): figure width, defaults to 800</span>
<span class="sd">            height (int): figure height, defaults to 2000</span>
<span class="sd">            cell_lines (Optional[List[str]]): set of cell lines to analyze,</span>
<span class="sd">            defaults to all cell lines.</span>
<span class="sd">            data_type (str): type of data to plot.</span>
<span class="sd">                            Can be &#39;means&#39;, &#39;modes&#39;, &#39;diffs&#39;, or &#39;pvalue&#39;.</span>
<span class="sd">            backend (Optional[str]): plotting backend to override default</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: invalid data type</span>
<span class="sd">            ValueError: invalid plotting backend</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">title_type</span><span class="p">:</span> <span class="nb">str</span>
        <span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;means&quot;</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">means</span>
            <span class="n">title_type</span> <span class="o">=</span> <span class="s2">&quot;Mean&quot;</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;modes&quot;</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">modes</span>
            <span class="n">title_type</span> <span class="o">=</span> <span class="s2">&quot;Mode&quot;</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;diffs&quot;</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">diffs</span>
            <span class="n">title_type</span> <span class="o">=</span> <span class="s2">&quot;Difference&quot;</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;pvalue&quot;</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">allelic_data</span>
            <span class="n">title_type</span> <span class="o">=</span> <span class="s2">&quot;p-value&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid data type&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">cell_lines</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cell_lines</span><span class="p">)]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">backend</span><span class="p">:</span>
            <span class="n">backend</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">viz_backend</span>

        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;plotly&quot;</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">:</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_create_heatmap</span><span class="p">(</span>
                <span class="n">data</span><span class="p">,</span> <span class="n">min_values</span><span class="p">,</span> <span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">,</span> <span class="n">title_type</span><span class="p">,</span> <span class="n">backend</span>
            <span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>

        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;matplotlib&quot;</span><span class="p">:</span>
            <span class="n">_</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_create_heatmap</span><span class="p">(</span>
                <span class="n">data</span><span class="p">,</span> <span class="n">min_values</span><span class="p">,</span> <span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">,</span> <span class="n">title_type</span><span class="p">,</span> <span class="n">backend</span>
            <span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>

        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid plotting backend&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicPositionData.sig_methylation_differences"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.sig_methylation_differences">[docs]</a>    <span class="k">def</span> <span class="nf">sig_methylation_differences</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">cell_lines</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">backend</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Display a graph figure showing a bar chart of significantly different</span>
<span class="sd">        mean / mode methylation across all or a subset of cell lines.</span>

<span class="sd">        Args:</span>
<span class="sd">            cell_lines (Optional[List[str]]): set of cell lines to analyze,</span>
<span class="sd">                                            defaults to all cell lines.</span>
<span class="sd">            backend (Optional[str]): plotting backend to override default</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: invalid plotting backend</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summarize_allelic_data</span><span class="p">(</span><span class="n">cell_lines</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">backend</span><span class="p">:</span>
            <span class="n">backend</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">viz_backend</span>

        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;plotly&quot;</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">:</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_create_methylation_diffs_bar_graph</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">backend</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>
        <span class="k">if</span> <span class="n">backend</span> <span class="o">==</span> <span class="s2">&quot;matplotlib&quot;</span><span class="p">:</span>
            <span class="n">_</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">Figure</span> <span class="o">=</span> <span class="n">viz</span><span class="o">.</span><span class="n">_create_methylation_diffs_bar_graph</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">backend</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">return</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid plotting backend&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicPositionData.generate_ad_stats"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.generate_ad_stats">[docs]</a>    <span class="k">def</span> <span class="nf">generate_ad_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Generate Anderson-Darling normality statistics for an individual data df.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: df of a-d test statistics</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="n">divide</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">invalid</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>  <span class="c1"># ignore divide-by-zero errors</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">individual_data</span>
        <span class="n">df2</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">applymap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_anderson_darling_test</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df2</span></div>

<div class="viewcode-block" id="GenomicPositionData.summarize_allelic_data"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.GenomicPositionData.summarize_allelic_data">[docs]</a>    <span class="k">def</span> <span class="nf">summarize_allelic_data</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">cell_lines</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Create a dataframe only of likely allelic methylation positions.</span>

<span class="sd">        Args:</span>
<span class="sd">            cell_lines (Optional[List[str]]): set of cell lines to analyze,</span>
<span class="sd">            defaults to all cell lines.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: dataframe of cell lines with likely allelic positions</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">cell_lines</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">allelic_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">allelic_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cell_lines</span><span class="p">)]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">allelic_data</span>

        <span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="n">divide</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">invalid</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>  <span class="c1"># ignore divide-by-zero errors</span>
        <span class="n">sig_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">ArrayLike</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;cellLine&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;ad_stat&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;p_crit&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;raw&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">row</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">column</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">value</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
                    <span class="n">sig_dict</span><span class="p">[</span><span class="s2">&quot;cellLine&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
                    <span class="n">sig_dict</span><span class="p">[</span><span class="s2">&quot;position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">column</span><span class="p">)</span>
                    <span class="n">sig_dict</span><span class="p">[</span><span class="s2">&quot;raw&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
                    <span class="n">sig_dict</span><span class="p">[</span><span class="s2">&quot;ad_stat&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
                    <span class="n">sig_dict</span><span class="p">[</span><span class="s2">&quot;p_crit&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">sig_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">sig_dict</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sig_df</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_anderson_darling_test</span><span class="p">(</span><span class="n">raw_list</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">AD_stats</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Run the Anderson-Darling normality test on methylation data at a point.</span>

<span class="sd">        Args:</span>
<span class="sd">            raw_list (pd.Series): list of fractional methylation levels per read.</span>

<span class="sd">        Returns:</span>
<span class="sd">            AD_stats: is the data significantly</span>
<span class="sd">                allelic (bool), A-D statistic (float), critical values (list)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">raw_list</span><span class="p">)):</span>
            <span class="n">stat</span><span class="p">:</span> <span class="n">ArrayLike</span>
            <span class="n">crits</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">ArrayLike</span><span class="p">]</span>
            <span class="n">stat</span><span class="p">,</span> <span class="n">crits</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">anderson</span><span class="p">(</span><span class="n">raw_list</span><span class="p">)</span>
            <span class="n">is_sig</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="nb">bool</span><span class="p">(</span><span class="n">stat</span> <span class="o">&gt;</span> <span class="n">crits</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>  <span class="c1"># type: ignore</span>
            <span class="k">return</span> <span class="n">AD_stats</span><span class="p">(</span><span class="n">is_sig</span><span class="p">,</span> <span class="n">stat</span><span class="p">,</span> <span class="n">crits</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">AD_stats</span><span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">])</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_chisquared_test</span><span class="p">(</span>
        <span class="n">data_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Perform Chi-squared analysis of a set of methylation calls.</span>

<span class="sd">        Args:</span>
<span class="sd">            data_list (List[int]): list of methylated (1)</span>
<span class="sd">                                    and unmethylated (0) values for a read.</span>
<span class="sd">            cutoff (float): pvalue cuttoff, defaults to 0.001</span>

<span class="sd">        Returns:</span>
<span class="sd">            Optional[float]: Chi-squared p-value if below cutoff, or None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">data_list</span><span class="p">)):</span>
            <span class="n">my_data</span><span class="p">:</span> <span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">int_</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="p">[</span><span class="n">data_list</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">data_list</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="mi">0</span><span class="p">)]</span>
            <span class="p">)</span>
            <span class="n">pvalue</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">chisquare</span><span class="p">(</span><span class="n">my_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">pvalue</span> <span class="o">&lt;=</span> <span class="n">cutoff</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">pvalue</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_return_individual_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell_line</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Return a dataframe of methylation at genomic position resolution.</span>

<span class="sd">        Args:</span>
<span class="sd">            cell_line (str): Cell line to analyze.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: dataframe of methylation at genomic position resolution.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">_find_positions</span><span class="p">(</span>
            <span class="n">pos</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">data_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">quma</span><span class="o">.</span><span class="n">Reference</span><span class="p">],</span> <span class="n">gseq</span><span class="p">:</span> <span class="nb">str</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]:</span>
            <span class="sd">&quot;&quot;&quot;Find individual resolution positions and data in quma results.</span>

<span class="sd">            Args:</span>
<span class="sd">                pos (str): position of read.</span>
<span class="sd">                data_list (List[quma.Reference]): list of quma read data</span>
<span class="sd">                gseq (str): reference genomic sequence of read position</span>

<span class="sd">            Returns:</span>
<span class="sd">                List[Tuple[List[int], List[str]]]: list of data from quma results.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">output</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">each</span> <span class="ow">in</span> <span class="n">data_list</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">offset</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="p">[</span>
                        <span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()</span>
                        <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">each</span><span class="o">.</span><span class="n">res</span><span class="o">.</span><span class="n">gAli</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="n">gseq</span><span class="p">)</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="c1"># FIXME: This only matches initial substring in cases of gaps</span>
                <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>

                <span class="n">pos_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span> <span class="o">+</span> <span class="n">offset</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s2">&quot;CG&quot;</span><span class="p">,</span> <span class="n">each</span><span class="o">.</span><span class="n">res</span><span class="o">.</span><span class="n">gAli</span><span class="p">)</span>
                <span class="p">]</span>

                <span class="n">my_point</span> <span class="o">=</span> <span class="p">(</span><span class="n">pos_list</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">each</span><span class="o">.</span><span class="n">res</span><span class="o">.</span><span class="n">val</span><span class="p">))</span>
                <span class="n">output</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">my_point</span><span class="p">)</span>

            <span class="k">return</span> <span class="n">output</span>

        <span class="k">def</span> <span class="nf">_collate_positions</span><span class="p">(</span>
            <span class="n">output</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
            <span class="sd">&quot;&quot;&quot;Collect together list of quma data points into position-indexed dict.</span>

<span class="sd">            Args:</span>
<span class="sd">                output (List[Tuple[List[int], List[str]]]): list of read data points</span>

<span class="sd">            Returns:</span>
<span class="sd">                Dict[int, str]: position-indexed genomic resolution methylation dict.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">result_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">each</span> <span class="ow">in</span> <span class="n">output</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">each</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">my_val</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">each</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">idx</span><span class="p">]</span>
                    <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>  <span class="c1"># pragma: no cover</span>
                        <span class="k">pass</span>
                    <span class="k">if</span> <span class="n">my_val</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="n">result_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">val</span><span class="p">):</span>
                            <span class="n">result_dict</span><span class="p">[</span><span class="n">val</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">my_val</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">result_dict</span><span class="p">[</span><span class="n">val</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">my_val</span><span class="p">]</span>

            <span class="n">result_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">result_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
            <span class="k">return</span> <span class="n">result_dict</span>

        <span class="k">def</span> <span class="nf">_make_pos_df</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
            <span class="sd">&quot;&quot;&quot;Convert dictionary of genomic resolution data into dataframe.</span>

<span class="sd">            Args:</span>
<span class="sd">                data (Dict[int, List[str]]): pos-indexed high res. methylation dict.</span>

<span class="sd">            Returns:</span>
<span class="sd">                pd.DataFrame: dataframe of pos-indexed high resolution methylation.</span>
<span class="sd">            &quot;&quot;&quot;</span>

            <span class="n">int_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">int_df</span>

        <span class="n">query</span><span class="p">:</span> <span class="n">QumaResult</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">quma_results</span><span class="p">[</span><span class="n">cell_line</span><span class="p">]</span>
        <span class="n">positions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">_positions</span>
        <span class="n">data</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">quma</span><span class="o">.</span><span class="n">Quma</span><span class="p">]</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">quma_output</span>
        <span class="n">big_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pos</span><span class="p">,</span> <span class="n">dat</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
            <span class="n">big_list</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">_find_positions</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">dat</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="n">dat</span><span class="o">.</span><span class="n">_gseq</span><span class="p">))</span>

        <span class="n">full_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">_collate_positions</span><span class="p">(</span><span class="n">big_list</span><span class="p">)</span>

        <span class="n">final_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">_make_pos_df</span><span class="p">(</span><span class="n">full_data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">final_df</span></div>


<span class="c1">##################################################################################</span>
<span class="c1">##################################################################################</span>

<span class="c1"># Module level helper functions</span>


<div class="viewcode-block" id="configure"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.configure">[docs]</a><span class="k">def</span> <span class="nf">configure</span><span class="p">(</span>
    <span class="n">base_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">prom_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">prom_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">prom_end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">offset</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">test_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">fname_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">viz_backend</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">results_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Config</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Helper method to set up all our environmental variables, such as for testing.</span>

<span class="sd">    Args:</span>
<span class="sd">        base_path (str): directory where all processing will occur, put .bam files</span>
<span class="sd">                         in &quot;test&quot; sub-directory in this folder</span>
<span class="sd">        prom_file (str): filename of genmic sequence of promoter region of interest</span>
<span class="sd">        prom_start (int): start position to analyze in promoter region</span>
<span class="sd">        prom_end (int): final position to analyze in promoter region</span>
<span class="sd">        chrom (str): chromosome promoter is located on</span>
<span class="sd">        offset (int): genomic position of promoter to offset reads</span>
<span class="sd">        test_dir(Optional[str]): name of test directory where bam files are located</span>
<span class="sd">        fname_pattern(Optional[str]): regex pattern for processing filenames</span>
<span class="sd">        viz_backend (Optional[str]): which plotting backend to use</span>
<span class="sd">        results_dir (Optional[str]): name of results directory</span>

<span class="sd">    Returns:</span>
<span class="sd">        Config: configuration dataclass instance.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">test_dir</span><span class="p">:</span>
        <span class="n">test_dir</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span>

    <span class="n">fname_pattern_compiled</span><span class="p">:</span> <span class="n">Pattern</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">fname_pattern</span><span class="p">:</span>
        <span class="n">fname_pattern_compiled</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^[a-zA-Z]+_([a-zA-Z0-9]+)_.+bam$&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fname_pattern_compiled</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">fname_pattern</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">viz_backend</span><span class="p">:</span>
        <span class="n">viz_backend</span> <span class="o">=</span> <span class="s2">&quot;plotly&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">results_dir</span><span class="p">:</span>
        <span class="n">results_dir</span> <span class="o">=</span> <span class="s2">&quot;results&quot;</span>

    <span class="n">config</span> <span class="o">=</span> <span class="n">Config</span><span class="p">(</span>
        <span class="n">base_directory</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">base_path</span><span class="p">),</span>
        <span class="n">promoter_file</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span> <span class="o">/</span> <span class="n">prom_file</span><span class="p">,</span>
        <span class="n">results_directory</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span> <span class="o">/</span> <span class="n">results_dir</span><span class="p">,</span>
        <span class="n">analysis_directory</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span> <span class="o">/</span> <span class="n">test_dir</span><span class="p">,</span>
        <span class="n">promoter_start</span><span class="o">=</span><span class="n">prom_start</span><span class="p">,</span>
        <span class="n">promoter_end</span><span class="o">=</span><span class="n">prom_end</span><span class="p">,</span>
        <span class="n">chromosome</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
        <span class="n">offset</span><span class="o">=</span><span class="n">offset</span><span class="p">,</span>
        <span class="n">viz_backend</span><span class="o">=</span><span class="n">viz_backend</span><span class="p">,</span>
        <span class="n">fname_pattern</span><span class="o">=</span><span class="n">fname_pattern_compiled</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">config</span></div>


<div class="viewcode-block" id="make_list_of_bam_files"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.make_list_of_bam_files">[docs]</a><span class="k">def</span> <span class="nf">make_list_of_bam_files</span><span class="p">(</span><span class="n">config</span><span class="p">:</span> <span class="n">Config</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;Check analysis directory for all valid .bam files.</span>

<span class="sd">    Args:</span>
<span class="sd">        config (Config): pyllelic configuration options.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list[str]: list of files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">config</span><span class="o">.</span><span class="n">analysis_directory</span><span class="o">.</span><span class="n">iterdir</span><span class="p">()</span> <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s2">&quot;.bam&quot;</span><span class="p">]</span></div>


<div class="viewcode-block" id="pyllelic"><a class="viewcode-back" href="../../pyllelic.html#pyllelic.pyllelic.pyllelic">[docs]</a><span class="k">def</span> <span class="nf">pyllelic</span><span class="p">(</span><span class="n">config</span><span class="p">:</span> <span class="n">Config</span><span class="p">,</span> <span class="n">files_set</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">GenomicPositionData</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Wrapper to call pyllelic routines.</span>

<span class="sd">    Args:</span>
<span class="sd">        config (Config): pyllelic config object.</span>
<span class="sd">        files_set (List[str]): list of bam files to analyze.</span>

<span class="sd">    Returns:</span>
<span class="sd">        GenomicPositionData: GenomicPositionData pyllelic object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">GenomicPositionData</span><span class="p">(</span><span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span> <span class="n">files_set</span><span class="o">=</span><span class="n">files_set</span><span class="p">)</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Andrew J. Bonham.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>